### PR TITLE
Add UnprocessableEntityResult, UnprocessableEntityObjectResult and ControllerBase.UnprocessableEntity methods

### LINKED ISSUE #6795: Consider adding UnprocessableEntityResult and ControllerBase.UnprocessableEntity
`422 Unprocessable Entity` is commonly used for semantic validation errors in APIs, as opposed to `400 Bad Request` for syntactical errors:

> The 422 (Unprocessable Entity) status code means the server
understands the content type of the request entity (hence a
415(Unsupported Media Type) status code is inappropriate), and the
syntax of the request entity is correct (thus a 400 (Bad Request)
status code is inappropriate) but was unable to process the contained
instructions. For example, this error condition may occur if an XML
request body contains well-formed (i.e., syntactically correct), but
semantically erroneous, XML instructions.

It would be nice if MVC included an `UnprocessableEntityResult` and an accompanying `ControllerBase.UnprocessableEntity` method.

They'd basically mirror what `BadRequestResult` and `ControllerBase.BadRequest` does, only with a different status code.

This has been [requested before](https://github.com/aspnet/Mvc/issues/1378#issuecomment-66729387) by @ashmind, but was never followed up.

### PR DESCRIPTION
**Pull request by khellang**
Closes #6795

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/UnprocessableEntityObjectResult.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        /// <param name="modelState"><see cref="ModelStateDictionary"/> containing the validation errors.</param>
+        public UnprocessableEntityObjectResult(ModelStateDictionary modelState)
+            : this(new SerializableError(modelState))
+        {
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="UnprocessableEntityObjectResult"/> instance.
+        /// </summary>
+        /// <param name="error">Contains the errors to be returned to the client.</param>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/UnprocessableEntityObjectResult.cs]: Contains ~~the~~ errors

File: test/Microsoft.AspNetCore.Mvc.Core.Test/UnprocessableEntityObjectResultTests.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            // Assert
+            Assert.Equal(StatusCodes.Status422UnprocessableEntity, result.StatusCode);
+            Assert.Equal(obj, result.Value);
+        }
+
+        [Fact]
+        public void UnprocessableEntityObjectResult_ModelState_SetsStatusCodeAndValue()
+        {
+            // Arrange & Act
+            var result = new UnprocessableEntityObjectResult(new ModelStateDictionary());
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/UnprocessableEntityObjectResultTests.cs]: Maybe toss in an error in to the MSD so we know it's the type that's being serialized.

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 3]
```diff
@@ -1499,6 +1499,41 @@ public virtual BadRequestObjectResult BadRequest(ModelStateDictionary modelState
         }
 
         /// <summary>
+        /// Creates an <see cref="UnprocessableEntityResult"/> that produces a <see cref="StatusCodes.Status422UnprocessableEntity"/> response.
+        /// </summary>
+        /// <returns>The created <see cref="UnprocessableEntityResult"/> for the response.</returns>
+        [NonAction]
+        public virtual UnprocessableEntityResult UnprocessableEntity()
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: `public virtual UnprocessableEntityResult UnprocessableEntity() => new UnprocessableEntityResult()` for 💯 points

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 4]
```diff
@@ -1499,6 +1499,41 @@ public virtual BadRequestObjectResult BadRequest(ModelStateDictionary modelState
...
+        {
+            return new UnprocessableEntityResult();
+        }
+
+        /// <summary>
+        /// Creates an <see cref="UnprocessableEntityObjectResult"/> that produces a <see cref="StatusCodes.Status422UnprocessableEntity"/> response.
+        /// </summary>
+        /// <returns>The created <see cref="UnprocessableEntityObjectResult"/> for the response.</returns>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: Needs a `<param >`

File: test/Microsoft.AspNetCore.Mvc.Core.Test/UnprocessableEntityObjectResultTests.cs (hunk 2 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Mvc.Core.Test/UnprocessableEntityObjectResultTests.cs]: There's no serialization in the test, and taht's fine. We don't need to overthink it

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs (hunk 3 above)
Comment (khellang) [reply src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: There are many other methods like this in the same class, it it worth the inconsistency?

Comment (khellang) [reply src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: Let's leave this for #6864 :smile:

**Comment by pranavkm** [comment]
Thanks for the PR!

**Pull request merged**